src/dft.py

```diff
@@ -1,14 +1,71 @@
 """A module for depth-first (in-order) traversal of trees."""
 
+from dataclasses import dataclass
 from typing import Iterable
 from tree import T
 
 
+@dataclass
+class stack:
+    stack: list[None | T | int]
+
+    def add(self, element: T | None | int) -> None:
+        self.stack.append(element)
+
+    def top(self) -> T | int | None:
+        return self.stack[-1]
+
+    def pop(self) -> T | int | None:
+        return self.stack.pop()
+
+    def is_empty(self) -> bool:
```

Comment:
It is more common to have loops that run while a data structure is *not* empty, rather than while they *are* empty, and while I fully acknowledge that most of the literature use loops such as

```python
while not stack.is_empty():
   ...
```

the typical Python idiom is

```python
while stack:
   ...
```

If you want to use this idiot, implement a `__bool__()` method that returns `True` when the stack is not empty and `False` otherwise.

```python
    def __ bool___(self) -> bool:
        return bool(self.stack)
```